File: BillingSystem.Sales/Domain/ValueObjects/Money.cs
namespace BillingSystem.Sales.Domain.ValueObjects;

public class Money
{
    public decimal Value { get; }

    public Money(decimal value)
    {
        if (value < 0)
            throw new ArgumentException("Money value cannot be negative.");
        Value = Math.Round(value, 4, MidpointRounding.ToEven);
    }

    public static Money operator +(Money m1, Money m2)
    {
        return new Money(m1.Value + m2.Value);
    }

    public static Money operator -(Money m1, Money m2)
    {
        return new Money(m1.Value - m2.Value);
    }

    public static Money operator *(Money m, decimal factor)
    {
        return new Money(m.Value * factor);
    }

    public static Money operator /(Money m, decimal divisor)
    {
        return new Money(m.Value / divisor);
    }

    public Money ApplyDiscountPercentage(Percentage discountPercentage)
    {
        decimal discountAmount = Value * (discountPercentage.Value / 100);
        return new Money(discountAmount);
    }
    public Money ApplyDiscountAmount(Money discountAmount)
    {
        return new Money(Value - discountAmount.Value);
    }

    public override string ToString() => Value.ToString("0.0000");
}

File: BillingSystem.Sales/Client/src/app/models/Money.ts
import { Percentage } from "./Percentage";

export class Money {
    public value: number;

    constructor(value: number) {
        if (value < 0) throw new Error("Money value cannot be negative.");
        this.value = parseFloat(value.toFixed(4));
    }

    public static add(m1: Money, m2: Money): Money {
        return new Money(m1.value + m2.value);
    }

    public static subtract(m1: Money, m2: Money): Money {
        return new Money(m1.value - m2.value);
    }

    public static multiply(m: Money, factor: number): Money {
        return new Money(m.value * factor);
    }

    public static divide(m: Money, divisor: number): Money {
        return new Money(m.value / divisor);
    }

    public applyDiscount(discountPercentage: Percentage): Money {
        const discountAmount = this.value * (discountPercentage.value / 100);
        return new Money(discountAmount);
    }

    public applyDiscountAmount(discountAmount: Money){
        return new Money(this.value - discountAmount.value);
    }
}
